### ISSUE TITLE
False positive TypeDoesNotContainType introduced recently (4.15 maybe)

### ISSUE THREAD

**Issue by VincentLanglet** [vimeo/psalm#7109]
An example can be found with this PR: https://github.com/sonata-project/SonataAdminBundle/pull/7643

Code like
```
$class = \stdClass::class;
static::assertSame([$class => ['sonata.user.admin.group1']], $pool->getAdminClasses());
```
is reported 
```
TypeDoesNotContainType: Cannot resolve types for $pool->getadminclasses() - array<class-string, array<array-key, string>> does not contain array{stdClass: array{"sonata.user.admin.group1"}}
```
But to me `array<class-string, array<array-key, string>>` contains `array{stdClass: array{"sonata.user.admin.group1"}}`.

**Comment by orklah** [comment]
Psalm seem to reduce the class-string in the offset to a simple string, and then complains that the string is not a class-string. I'm not sure what could have triggered this in recent versions.

Are you able to create a reproducer on psalm.dev?

**Comment by VincentLanglet** [comment]
> Psalm seem to reduce the class-string in the offset to a simple string, and then complains that the string is not a class-string. I'm not sure what could have triggered this in recent versions.

Like 2-3 days ago there was no error, so it's pretty recent.

> Are you able to create a reproducer on psalm.dev?

There is no error in https://psalm.dev/r/e62634509f 
I'm not sure if it's also related to the assertSame then.

**Comment by VincentLanglet** [comment]
@orklah Changing the code to
```
static::assertTrue([$class => ['sonata.user.admin.group1']] === $pool->getAdminClasses());
```
solve the issue. so I don't know what's the weird interaction with `assertSame`.

**Comment by orklah** [comment]
Can be reproduced by https://psalm.dev/r/1ed5e9e664

Seems to be caused by https://github.com/vimeo/psalm/pull/7076

@klimick mind taking a look?

**Comment by klimick** [comment]
Before #7076 next code was valid:

```php
class A
{
    public function test(): void
    {
        // array{stdClass: ""}
        $a = [stdClass::class => ''];
        // array{Traversable: ""}
        $b = [Traversable::class => ''];

        $this->assertSame($a, $b);
    }

    /**
     * @template T
     *
     * @param T $expected
     * @param mixed $actual
     * @psalm-assert =T $actual
     */
    public function assertSame($expected, $actual): void
    {
        return;
    }
}
```

But is it really valid?
Psalm considered it valid because he compared `array` with `array` before.

Now Psalm tries to compare `array{stdClass: ''}` with `array<class-string, mixed>`.
Where `stdClass` is just `string`.
Not `class-string`.

As I know `TKeyedArray` can't have type different from `int | string` for keys.
For this reason `stdClass` is string.

@orklah what should we do?

**Comment by orklah** [comment]
Oh, nice so this isn't an issue with the assertion system itself.

> As I know TKeyedArray can't have type different from int | string for keys.

subtypes of those two works as well, for example, you can have `"hello"` as an offset, which is a TLiteralString, so TClassString should be no exception

So it seems like Psalm resolves `stdClass::class` to `'stdClass'` but it shouldn't, so this is what has to change

**Comment by klimick** [comment]
> So it seems like Psalm resolves `stdClass::class` to `'stdClass'` but it shouldn't, so this is what has to change

Psalm already preserves info about `class-string` in TKeyedArray (the `$class_strings` property), but does not use in any way.
We could add the output of this info to the `getId` method but Psalm don't understand types like `array{stdClass::class: mixed}` at the parser level currently.

For me, extending the parser seems like a bit of a challenge)

**Comment by weirdan** [comment]
It doesn't have to touch the parser part, if we accept the ambiguity of `$a = [stdClass::class => 1]` vs `$a = ['stdClass' => 1]`. Both of those could be represented by `array{stdClass: 1}`, but they may differ internally. 

Wouldn't be the first time we do this either - e.g. there are a quite a few flags on `Union` that do not affect textual type representation.

**Comment by klimick** [comment]
> It doesn't have to touch the parser part, if we accept the ambiguity of `$a = [stdClass::class => 1]` vs `$a = ['stdClass' => 1]`. Both of those could be represented by `array{stdClass: 1}`, but they may differ internally.

Sorry. I don't understand how it can help.
`$a = [stdClass::class => 1]` and `$a = ['stdClass' => 1]` can be compared textually, but `array<class-string, int>` and `array{stdClass: 0}` not.

**Comment by orklah** [comment]
If the class-string is actually preserved in the TKeyedArray, then the fault must be in the comparison of the two types

**Comment by klimick** [comment]
> If the class-string is actually preserved in the TKeyedArray, then the fault must be in the comparison of the two types

Before the actual type is compared, it will be serialized to a string with the `getId`:
https://github.com/vimeo/psalm/blob/dee3fc43587e88c1d100adf40252660f949ec4b7/src/Psalm/Internal/Analyzer/Statements/Expression/CallAnalyzer.php#L789

Information about `class-string` will be lost later in the AssertionReconciler. During the type parsing:
https://github.com/vimeo/psalm/blob/39402c233d4856447615dfccf1d3431eca6ca6c9/src/Psalm/Internal/Type/AssertionReconciler.php#L341

Why problem not with `TKeyedArray` serialization/deserialization?)

**Comment by orklah** [comment]
Oh, yeah, you're right, I missed the fact that it was serialized here.

I'd go for changing the way we display class-strings in keyed array then. One of those two syntaxes: https://psalm.dev/r/f780ccaa84

**Comment by weirdan** [comment]
It would be interesting to explore what it would take to avoid type -> text -> type roundtrip though.

**Comment by orklah** [comment]
Yeah, I thought about that too, but going full `serialize($union)` is a bit much in my taste. Also, that's a hole in the syntax not to allow class-strings in a keyed array. Some users may want to do that (as a param type for example)

**Comment by weirdan** [comment]
Re proposed syntax options: both are used in Psalm type system, so they not necessarily alternatives: https://psalm.dev/r/6e690bd75f

**Comment by weirdan** [comment]
So for keyed array, where key names have to be known, `stdClass::class` syntax is the only one that makes sense.

**Comment by orklah** [comment]
yeah, you're right, `::class` syntax is actually a `TDependentGetClass` if I'm not mistaken. It's used to say it could actually be a class-string of a child of shown class I believe

**Comment by weirdan** [comment]
> `::class` syntax is actually a `TDependentGetClass` if I'm not mistaken

It's `TLiteralClassString`

**Comment by VincentLanglet** [comment]
I'm not sure to understand. Is there a "quick" solution or should https://github.com/vimeo/psalm/pull/7076 be reverted ?

It's better to have false negative like https://github.com/vimeo/psalm/issues/7109#issuecomment-990070219 than having false positive.

**Comment by orklah** [comment]
I tried to look at it yesterday, it seems the class-string doesn't make it to the TKeyedArray. I'll have to check what can be done about that.

Ideally, I'd like to avoid reverting because the change is legit. We'll try to get it fixed for the next release, in the meantime, feel free to go back to the previous version

**Comment by orklah** [comment]
Ok, the class-string do make it to the TKeyedArray, I was not expecting it to be stored like that. I'm able to change the output, I'm now trying to check the parser
